optimade/models/structures.py

```diff
@@ -758,6 +764,62 @@ class StructureResourceAttributes(EntryResourceAttributes):
         queryable=SupportLevel.MUST,
     )
 
+    @validator("chemical_formula_reduced", "chemical_formula_hill")
+    def check_ordered_formula(cls, v, field):
+        if v is None:
+            return v
+
+        numbers = re.findall(r"(\d+\.?)+", v)
+        elements = re.findall(r"[A-Z][a-z]?", v)
+        expected_elements = sorted(elements)
+
+        if field.name == "chemical_formula_hill":
+            for elem in ("C", "H"):
+                if elem in expected_elements:
+                    expected_elements.pop(expected_elements.index(elem))
+                    expected_elements.insert(0, elem)
+
+        if any(elem not in CHEMICAL_SYMBOLS for elem in elements):
+            raise ValueError(
+                f"Cannot use unknown chemical symbols {[elem for elem in elements if elem not in CHEMICAL_SYMBOLS]} in {field.name!r}"
+            )
+
+        if "1" in numbers:
+            raise ValueError(
+                "Must omit proportion '1' from formula {v} in {field.name!r}"
+            )
+        if expected_elements != elements:
+            order = "Hill" if field == "chemical_formula_hill" else "alphabetical"
+            raise ValueError(
+                f"Elements in {field.name!r} must appear in {order} order: {expected_elements} not {elements}."
+            )
+
+        return v
+
+    @validator("chemical_formula_anonymous")
+    def check_anonymous_formula(cls, v):
+        elements = tuple(re.findall(r"[A-Z][a-z]*", v))
+        numbers = [int(n.strip()) for n in re.split(r"[A-Z][a-z]*", v)[1:] if n.strip()]
```

Comment:
Do you need to avoid the first entry if you're already checking `n` with `n.strip()`?
```suggestion
        numbers = [int(n.strip()) for n in re.split(r"[A-Z][a-z]*", v) if n.strip()]
```

Reply:
Fair enough, will add this to the second commit